packages/world/src/systems/MineSystem.sol

```diff
@@ -222,50 +232,67 @@ contract MineSystem is System {
     }
   }
 
-  // TODO: this is ugly, but doing this to avoid stack too deep errors. We should refactor later.
-  function _processMassReduction(EntityId caller, Vec3 callerCoord, uint16 toolSlot, EntityId mined)
-    internal
-    returns (uint128, bool)
-  {
+  function _applyMassReduction(
+    EntityId caller,
+    Vec3 callerCoord,
+    uint16 toolSlot,
+    ObjectType minedType,
+    uint128 massLeft
+  ) internal returns (uint128, bool) {
+    if (massLeft == 0) {
+      return (0, true);
+    }
+
     ToolData memory toolData = InventoryUtils.getToolData(caller, toolSlot);
-    (uint128 finalMass, uint128 toolMassReduction, uint128 energyReduction) = _getMassReduction(toolData, mined);
+    uint128 toolMultiplier = _getToolMultiplier(toolData.toolType, minedType);
+
+    // round up since we want to use up to massLeft after applying the multiplier
+    uint128 toolMassReduction = toolData.getMassReduction(massLeft.divUp(toolMultiplier));
```

Comment:
could we not instead multiply the multiplier inside of `getMassReduction` and then you wouldnt need to divide and get rounding errors?

Reply:
did something similar in the end, did some refactoring to pass the multiplier to the tool and return both the tool and the total mass reduction. It wasn't as simple as just multiplying within the getMassReduction function because we need to figure out the tool mass reduction after applying the multiplier (basically we get the proportional amount of mass to reduce from the tool depending on how much mass we are able to reduce from massLeft)